File: src/test/java/at/george/wordcount/WordCounterTest.java
package at.george.wordcount;

import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;

import static org.junit.jupiter.api.Assertions.assertEquals;

class WordCounterTest {

    public static final String PATH_TO_STOPWORDS = "test_stopwords.txt";
    private WordCounter counter = new WordCounter();

    @ParameterizedTest
    @CsvSource({
            "'', 0",
            "word, 1",
            "word word, 2",
            "word1word word, 1",
            "'word word           wod', 3",
            "'. , 1 / *', 0",
            "'Foo1 Bar! Baz, Baz. word', 1",
            "' wordWithSpaceBefore', 1",
            "'wordWithSpaceAfter ', 1"
    })
    void count(final String text, final int expected) {
        int actual = counter.count(text);

        assertEquals(expected, actual);
    }

    @ParameterizedTest
    @CsvSource({
            "'', 0",
            "word, 1",
            "word word, 2",
            "word1word word, 1",
            "'word word           wod', 3",
            "'. , 1 / *', 0",
            "'Foo1 Bar! Baz, Baz. word', 1",
            "' wordWithSpaceBefore', 1",
            "'wordWithSpaceAfter ', 1",
            "the a word, 1",
    })
    void countWithStopWords(final String text, final int expected) {
        int actual = counter.count(text, PATH_TO_STOPWORDS);

        assertEquals(expected, actual);
    }
}

File: src/test/kotlin/at/george/wordcount/WordCounterTest.kt
package at.george.wordcount

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

class WordCounterTest {
    private val wordCounter = WordCounter()

    @Test
    fun `null text should return 0`() {
        val text = emptyList<String>()
        val result = wordCounter.count(text, emptyList()).numberOfWords

        assertEquals(0, result)
    }

    @Test
    fun `should return right number of words`() {
        val text = listOf("Marry", "had", "a", "little", "lamb")
        val result = wordCounter.count(text, emptyList()).numberOfWords

        assertEquals(5, result)
    }

    @Test
    fun `should not count the words passed as exceptions`() {
        val text = listOf("Marry", "had", "a", "little", "lamb", "the")
        val result = wordCounter.count(text, listOf("on", "off", "the", "a")).numberOfWords

        assertEquals(4, result)
    }

    @Test
    fun `should return the right number of unique words`() {
        val text = listOf("Marry", "had", "a", "little", "lamb", "Marry", "had", "a")
        val result = wordCounter.count(text, listOf("on", "off", "the", "a")).numberOfUniqueWords

        assertEquals(4, result)
    }
}